lte/gateway/c/oai/tasks/ha/ha_service_handler.cpp

```diff
@@ -0,0 +1,179 @@
+#include "s1ap_state_manager.h"
+
+static bool process_ue_context(
+    const hash_key_t keyP, void* const elementP, void* parameterP,
+    void** resultP);
+
+bool sync_up_with_orc8r(void) {
+  magma::HaClient::get_eNB_offload_state(
+      [](grpc::Status status,
+         magma::lte::GetEnodebOffloadStateResponse response) {
+        if (status.ok()) {
+          OAILOG_INFO(
+              LOG_UTIL, "Received eNodeB connection state with the primary.");
+          // iterate over the eNodeB connection states
+          ha_agw_offload_req_t offload_req = {0};
+          for (auto const& item : response.enodeb_offload_states()) {
+            if (item.second ==
+                magma::lte::GetEnodebOffloadStateResponse::PRIMARY_CONNECTED) {
+              offload_req.eNB_id = item.first;
+              // Offload any UE to check if it can be camped on the primary.
+              // The effect will be observed in the next sync up with the cloud.
+              offload_req.enb_offload_type = ANY;
+              handle_agw_offload_req(&offload_req);
+            } else if (
+                item.second == magma::lte::GetEnodebOffloadStateResponse::
+                                   PRIMARY_CONNECTED_AND_SERVING_UES) {
+              offload_req.eNB_id = item.first;
+              // Primary looks healthy as UEs are camped on it, offload the rest
+              // of UEs.
+              offload_req.enb_offload_type = ALL;
+              handle_agw_offload_req(&offload_req);
+            }
+          }
+        } else {
+          OAILOG_ERROR(
+              LOG_UTIL, "GRPC Failure Message: %s Status Error Code: %d",
+              status.error_message().c_str(), status.error_code());
+        }
+      });
+  return true;
+}
+
+typedef struct callback_data_s {
+  s1ap_state_t* s1ap_state;
+  ha_agw_offload_req_t* request;
+} callback_data_t;
+
+void handle_agw_offload_req(ha_agw_offload_req_t* offload_req) {
+  hash_table_ts_t* state_imsi_ht =
+      magma::lte::MmeNasStateManager::getInstance().get_ue_state_ht();
+  callback_data_t callback_data;
+  callback_data.s1ap_state =
+      magma::lte::S1apStateManager::getInstance().get_state(false);
+  callback_data.request = offload_req;
+  hashtable_ts_apply_callback_on_elements(
+      state_imsi_ht, process_ue_context, (void*) &callback_data, NULL);
+}
+
+bool process_ue_context(
+    const hash_key_t keyP, void* const elementP, void* parameterP,
+    void** resultP) {
+  imsi64_t imsi64                = INVALID_IMSI64;
+  callback_data_t* callback_data = (callback_data_t*) parameterP;
+  ha_agw_offload_req_t* offload_request =
+      (ha_agw_offload_req_t*) callback_data->request;
+  s1ap_state_t* s1ap_state = (s1ap_state_t*) callback_data->s1ap_state;
+  struct ue_mm_context_s* ue_context_p = (struct ue_mm_context_s*) elementP;
+  bool any_flag = false;  // true if we tried offloading any UE
+
+  IMSI_STRING_TO_IMSI64(offload_request->imsi, &imsi64);
+
+  enb_description_t* enb_ref_p =
```

Comment:
Wouldn't it be easier to store the  enb_id (in addition to sctp_assoc_id_key) in the ue_mm_context itself, than look up the s1ap_state hashtable for every UE at every AGW offload request?

Reply:
Indeed. Since the eNB state in the cloud is coming directly from s1ap state, it is better to keep consistency here. Cost-wise, since this is hash table look up and we do this sync up with the cloud every 5 minutes (and only at the secondary instance that has actual UEs camped on it), I think the processing costs are marginalized. However, in a separate PR, once we fix the inconsistency in terms of eNB id between s1ap and mme tasks, we can revisit this point.